lib/ansible/module_utils/basic.py

```diff
@@ -720,9 +720,16 @@ def warn(self, warning):
         warn(warning)
         self.log('[WARNING] %s' % warning)
 
-    def deprecate(self, msg, version=None):
-        deprecate(msg, version)
-        self.log('[DEPRECATION WARNING] %s %s' % (msg, version))
+    def deprecate(self, msg, version=None, date=None):
+        if version is not None and date is not None:
+            raise AssertionError("implementation error -- version and date must not both be set")
+        deprecate(msg, version=version, date=date)
+        # For compatibility, we accept that neither version nor date is set,
+        # and treat that the same as if version would haven been set
+        if date is not None:
+            self.log('[DEPRECATION WARNING] %s %s' % (msg, date))
+        else:
+            self.log('[DEPRECATION WARNING] %s %s' % (msg, version))
```

Comment:
instead of having 2 params allow version to be overloaded ?

Reply:
I would really prefer not to do that. Having two separate parameters makes it easier to validate, since it is clear from the parameter name what the contents must look like.

Reply:
i would at least make them mutually exclusive, really don't want people complaining about precedence of date ><>= version

Reply:
the main reason for using as overload is that we don't backport features, and this clearly is a feature.

Reply:
I agree we should keep 2 args so that we can validate in CI that they are formatted correctly. Making the mutually exclusive also sounds like a good plan.

Reply:
Indeed! And: lines 724 and 725 above raise `AssertionError("implementation error -- version and date must not both be set")` if both are set :)

Reply:
haha, was that already there? I didn't look, I just commented on a comment :)

Reply:
Hehe :) Github's excerpt is unfortunately hiding all the interesting parts and comments explaining these few lines ;)